### PR TITLE
(feat): integrate snapshot testing

### PR DESCRIPTION
**Pull request by AnthonyQuy**
PR to integrate snapshot testing #9989

### CODE REVIEW

File: src/matchers/general/toMatchSnapshot.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { toMatchSnapshot as jestToMatchSnapshot, toMatchInlineSnapshot as jestToMatchInlineSnapshot } from 'jest-snapshot'
```
Comment (christian-bromann) [inline src/matchers/general/toMatchSnapshot.ts:1]: We should define this as peer dependency

File: src/matchers/general/toMatchSnapshot.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+export async function toMatchSnapshot(received: any, propertiesOrHint?: object | string, hint?: string) {
```
Comment (christian-bromann) [inline src/matchers/general/toMatchSnapshot.ts:3]: `received` should be from type `WebdriverIO.Element` and within the matcher we should get the `outerHTML` structure of the element. A nice feature would be to allow to ignore certain properties or all properties at all.

**Review by christian-bromann — COMMENTED** [review]
This is a great first step. Let's make it useful for the WebdriverIO case.

File: src/matchers/general/toMatchSnapshot.ts (hunk 2 above)
Comment (AnthonyQuy) [reply src/matchers/general/toMatchSnapshot.ts:3]: I'm keeping the original jest.expect.toMatchSnapshot behavior here. For the element user can use like:

expect(element.outerHTML).toMatchSnapshot()

properly can implement expect(element).toMatchOuterHtmlSnapshot() later

File: src/matchers/general/toMatchSnapshot.ts (hunk 1 above)
Comment (AnthonyQuy) [reply src/matchers/general/toMatchSnapshot.ts:1]: noted with thanks!

File: src/matchers/general/toMatchSnapshot.ts (hunk 2 above)
Comment (christian-bromann) [reply src/matchers/general/toMatchSnapshot.ts:3]: Totally agree with that after reviewing the other PR. Keeping the `toMatchSnapshot` as is makes sense to avoid surprises with users. I would call the new matcher `toMatchElementSnapshot`.

File: src/matchers/general/toMatchSnapshot.ts (hunk 1 above)
Comment (AnthonyQuy) [reply src/matchers/general/toMatchSnapshot.ts:1]: updated

File: types/jest-global.d.ts [hunk 3]
```diff
@@ -1,7 +1,5 @@
 /// <reference types="./standalone.js"/>
 
-declare const expect: ExpectWebdriverIO.Expect
```
Comment (christian-bromann) [inline types/jest-global.d.ts:3]: What is the purpose of this removal? Seems unrelated to the changeset.

File: package.json [hunk 4]
```diff
@@ -85,8 +85,11 @@
     "webdriverio": "^8.23.1"
   },
   "dependencies": {
-    "expect": "^29.7.0",
+    "@jest/expect": "^29.7.0",
     "jest-matcher-utils": "^29.7.0",
     "lodash.isequal": "^4.5.0"
+  },
+  "peerDependencies": {
+    "jest-snapshot": "29.7.0"
```
Comment (christian-bromann) [inline package.json:93]: Since `@jest/expect` has this dependency defined with the same `^`
```suggestion
    "jest-snapshot": "^29.7.0"
```

File: types/jest-global.d.ts (hunk 3 above)
Comment (AnthonyQuy) [reply types/jest-global.d.ts:3]: image
this conflict again @types/jest/index.d.ts

Comment (AnthonyQuy) [reply types/jest-global.d.ts:3]: the ExpectWebdriverIO.expect has been updated in expect-webdriverio.d.ts file

**Comment by AnthonyQuy** [comment]
This PR should be ready to merge

File: package.json [hunk 5]
```diff
@@ -85,8 +85,11 @@
     "webdriverio": "^8.23.1"
   },
   "dependencies": {
-    "expect": "^29.7.0",
+    "@jest/expect": "^29.7.0",
     "jest-matcher-utils": "^29.7.0",
     "lodash.isequal": "^4.5.0"
+  },
+  "peerDependencies": {
+    "jest-snapshot": "^29.7.0"
```
[the remark this answers is not in the document]
Comment (AnthonyQuy) [reply package.json:93]: noted with thanks! I will update it

**Review by christian-bromann — APPROVED** [review]
Awesome, thanks 👍

**Comment by christian-bromann** [comment]
Hey @AnthonyQuy 👋  unfortunately I had to revert this PR. The problem is that Jest and its snapshot utilities make it really hard to port it into the browser. One very important use case is that we can run browser tests and therefor `expect` needs to work in the browser. With snapshotting there are a lot of `fs` operations going on, e.g. storing snapshots etc. I totally forgot to mention this when reviewing this PR.

There are a couple of options:
- look into `vitest` expect and see if it is more modular allowing us to determine whether we can to the snapshot assessment  in a custom place
- build our own

Note that in the browser case we can't load a snapshot file, we need to find a way to provide this capability in the browser.

**Comment by christian-bromann** [comment]
I created a new issue for this.
